### PR TITLE
Pkgoogle/efficient net migration

### PR DESCRIPTION
**Pull request by pkgoogle**
Migrating and consolidating EfficientNetV1 and V2 to keras-hub. 

PR for https://github.com/keras-team/keras-nlp/issues/1746

### CODE REVIEW

File: keras_nlp/src/layers/fusedmbconv.py [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+    ```
+    inputs = tf.random.normal(shape=(1, 64, 64, 32), dtype=tf.float32)
+    layer = keras_cv.layers.FusedMBConvBlock(
+        input_filters=32,
+        output_filters=32
+    )
+    output = layer(inputs)
+    output.shape # TensorShape([1, 224, 224, 48])
+    ```
+    """  # noqa: E501
```
Comment (mattdangerw) [inline keras_nlp/src/layers/fusedmbconv.py:86]: remove all these `noqa` markers.

File: keras_nlp/src/models/efficientnet_v1/efficientnet_v1_backbone_test.py [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+from absl.testing import parameterized
+
+from keras_nlp.src.models.efficientnet_v1.efficientnet_v1_backbone import (  # noqa: E501
+    EfficientNetV1Backbone,
+)
+from keras_nlp.src.tests.test_case import TestCase
+from keras_nlp.src.utils.keras_utils import get_feature_extractor
+
+
+class EfficientNetV1BackboneTest(TestCase):
```
Comment (mattdangerw) [inline keras_nlp/src/models/efficientnet_v1/efficientnet_v1_backbone_test.py:29]: Follow the form of other backbone tests that have been added. We have some common test routines.

File: keras_nlp/src/utils/keras_utils.py [hunk 3]
```diff
@@ -128,3 +128,101 @@ def standardize_data_format(data_format):
             f"Received: data_format={data_format}"
         )
     return data_format
+
+
+def get_feature_extractor(model, layer_names, output_keys=None):
```
Comment (mattdangerw) [inline keras_nlp/src/utils/keras_utils.py:133]: Let's remove this if we can and try to write tests without it.

File: keras_nlp/src/utils/keras_utils.py [hunk 4]
```diff
@@ -128,3 +128,101 @@ def standardize_data_format(data_format):
...
+    else:
+        if not keras.backend.is_keras_tensor(input_tensor):
+            return keras.layers.Input(
+                tensor=input_tensor, shape=input_shape, **kwargs
+            )
+        else:
+            return input_tensor
+
+
+def correct_pad_downsample(inputs, kernel_size):
```
Comment (mattdangerw) [inline keras_nlp/src/utils/keras_utils.py:206]: Is this something we will need generally? Why haven't any of our other backbones yet. Let's just add it to efficientnet if we are unsure if it's general.

**Review by mattdangerw — COMMENTED** [review]
Thanks for the PR!

- I think the first big question we should ask is if we want separate `EfficientNetV1` and `EfficientNetV2` classes. Keeping the separate will be more work, but could make sense if the architectures are very diverged.
- After that, it may make sense to land just one new public symbol at a time. Otherwise this could take a while to clean up and land.
- Let's make sure to rewrite for the "local style" of this repo. We are trying to make sure we aren't just doing a code dump, but actually come up with a more consistent and maintainable state than the keras-cv source.

File: keras_nlp/src/models/efficientnet_v1/efficientnet_v1_backbone.py [hunk 5]
```diff
@@ -0,0 +1,447 @@
...
+import keras
+
+from keras_nlp.src.api_export import keras_nlp_export
+from keras_nlp.src.utils import keras_utils
+from keras_nlp.src.models.backbone import Backbone
+
+
+@keras_nlp_export("keras_nlp.models.EfficientNetV1Backbone")
+@keras.saving.register_keras_serializable(package="keras_nlp.models")
+class EfficientNetV1Backbone(Backbone):
```
[the remark this answers is not in the document]
Comment (pkgoogle) [reply keras_nlp/src/models/efficientnet_v1/efficientnet_v1_backbone.py:25]: Sg, let me work on this first, as that might change the rest of this.

File: keras_nlp/src/models/efficientnet/mbconv.py [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+        kernel_size=3,
+        strides=1,
+        se_ratio=0.0,
+        bn_momentum=0.9,
+        activation="swish",
+        survival_probability: float = 0.8,
+        **kwargs
+    ):
+        """
+        Implementation of the MBConv block (Mobile Inverted Residual Bottleneck)
```
Comment (mattdangerw) [inline keras_nlp/src/models/efficientnet/mbconv.py:44]: same, one liner on the quote line, and keep it short

**Review by mattdangerw — COMMENTED** [review]
Looking good! Some API comments, haven't looked at test failures yet.

**Comment by mattdangerw** [comment]
@pkgoogle most of the test failures look easily fixable, convert list to tuple when comparing against a shape.

File: keras_nlp/src/models/efficientnet/efficientnet_backbone.py [hunk 7]
```diff
@@ -0,0 +1,577 @@
...
+            if use_depth_divisor_as_min_depth is set to True.
+        include_initial_padding: bool, whether to include initial zero padding
+            (as per v1).
+        use_depth_divisor_as_min_depth: bool, whether to use depth_divisor as
+            the minimum depth instead of min_depth (as per v1).
+        cap_round_filter_decrease: bool, whether to cap the max decrease in the
+            number of filters the rounding process potentially produces
+            (as per v1).
+        stem_conv_padding: str, can be 'same' or 'valid'. Padding for the stem.
+        bn_momentum: float, momentum for the moving average calcualtion in the
```
[the remark this answers is not in the document]
Comment (pkgoogle) [reply keras_nlp/src/models/efficientnet/efficientnet_backbone.py]: done

File: keras_nlp/src/models/efficientnet/efficientnet_backbone.py [hunk 8]
```diff
@@ -0,0 +1,577 @@
...
+        (https://github.com/keras-team/keras/blob/master/keras/applications/efficientnet.py)
+    - [EfficientNetV2: Smaller Models and Faster Training]
+        (https://arxiv.org/abs/2104.00298) (ICML 2021)
+
+    Args:
+        include_rescaling: bool, whether to rescale the inputs. If set to
+            True, inputs will be passed through a `Rescaling(1/255.0)` layer.
+        width_coefficient: float, scaling coefficient for network width.
+        depth_coefficient: float, scaling coefficient for network depth.
+        dropout_rate: float, dropout rate before final classifier layer.
```
[the remark this answers is not in the document]
Comment (pkgoogle) [reply keras_nlp/src/models/efficientnet/efficientnet_backbone.py]: done

File: keras_nlp/src/models/efficientnet/efficientnet_backbone.py [hunk 9]
```diff
@@ -0,0 +1,577 @@
...
+        drop_connect_rate: float, dropout rate at skip connections. The default
```
[the remark this answers is not in the document]
Comment (pkgoogle) [reply keras_nlp/src/models/efficientnet/efficientnet_backbone.py]: consolidated all the arguments into just "dropout"

File: keras_nlp/src/models/efficientnet/efficientnet_backbone.py [hunk 10]
```diff
@@ -0,0 +1,577 @@
...
+        stackwise_strides: list of ints, stackwise_strides for each conv block.
+        stackwise_squeeze_and_excite_ratios: list of ints, the squeeze and
+            excite ratios passed to the squeeze and excitation blocks.
+        stackwise_conv_types: list of strings.  Each value is either 'v1',
+            'unfused' or 'fused' depending on the desired blocks.  'v1' uses the
+            original efficientnet block. FusedMBConvBlock is similar to
+            MBConvBlock, but instead of using a depthwise convolution and a 1x1
+            output convolution blocks fused blocks use a single 3x3 convolution
+            block.
+        skip_connection_dropout: float, dropout rate at skip connections.
```
[the remark this answers is not in the document]
Comment (pkgoogle) [reply keras_nlp/src/models/efficientnet/efficientnet_backbone.py]: consolidated all the arguments into just "dropout"

File: keras_nlp/src/models/efficientnet/efficientnet_backbone.py [hunk 11]
```diff
@@ -0,0 +1,577 @@
...
+        stackwise_input_filters: list of ints, number of input filters for
+            each conv block.
+        stackwise_output_filters: list of ints, number of output filters for
+            each stack in the conv blocks model.
+        stackwise_expansion_ratios: list of floats, expand ratio passed to the
+            squeeze and excitation blocks.
+        stackwise_strides: list of ints, stackwise_strides for each conv block.
+        stackwise_squeeze_and_excite_ratios: list of ints, the squeeze and
+            excite ratios passed to the squeeze and excitation blocks.
+        stackwise_conv_types: list of strings.  Each value is either 'v1',
```
[the remark this answers is not in the document]
Comment (pkgoogle) [reply keras_nlp/src/models/efficientnet/efficientnet_backbone.py]: original EfficientNetV2 implementation had it per layer -- but I don't mind either way. Changed to "stackwise_block_types".

File: keras_nlp/src/models/efficientnet/efficientnet_backbone.py [hunk 12]
```diff
@@ -0,0 +1,577 @@
...
+                    x = self._apply_efficientnet_block(
+                        inputs=x,
+                        filters_in=input_filters,
+                        filters_out=output_filters,
+                        kernel_size=stackwise_kernel_sizes[i],
+                        strides=strides,
+                        expand_ratio=stackwise_expansion_ratios[i],
+                        se_ratio=squeeze_and_excite_ratio,
+                        activation=activation,
+                        dropout_rate=drop_connect_rate * block_id / blocks,
```
[the remark this answers is not in the document]
Comment (pkgoogle) [reply keras_nlp/src/models/efficientnet/efficientnet_backbone.py]: agreed, consolidated to "dropout" ... it seems the original implementation of those blocks had a bug ... they were using survival_probability as dropout -- but since the default was always set to 0.2 via this backbone it effectively never presented itself and acted appropriately.

File: keras_nlp/src/models/efficientnet/efficientnet_backbone.py [hunk 13]
```diff
@@ -0,0 +1,577 @@
...
+    )
+    images = np.ones((1, 256, 256, 3))
+    outputs = efficientnet.predict(images)
+    ```
+    """
+
+    def __init__(
+        self,
+        *,
+        include_rescaling,
```
[the remark this answers is not in the document]
Comment (pkgoogle) [reply keras_nlp/src/models/efficientnet/efficientnet_backbone.py]: done

File: keras_nlp/src/models/efficientnet/efficientnet_backbone.py [hunk 14]
```diff
@@ -0,0 +1,577 @@
...
+
+from keras_nlp.src.api_export import keras_nlp_export
+from keras_nlp.src.models.efficientnet.fusedmbconv import FusedMBConvBlock
+from keras_nlp.src.models.efficientnet.mbconv import MBConvBlock
+from keras_nlp.src.models.feature_pyramid_backbone import FeaturePyramidBackbone
+
+
+@keras_nlp_export("keras_nlp.models.EfficientNetBackbone")
+class EfficientNetBackbone(FeaturePyramidBackbone):
+    """Instantiates the EfficientNet architecture (encapsulates V1 & V2).
```
[the remark this answers is not in the document]
Comment (pkgoogle) [reply keras_nlp/src/models/efficientnet/efficientnet_backbone.py]: done

File: keras_nlp/src/models/efficientnet/efficientnet_backbone.py [hunk 15]
```diff
@@ -0,0 +1,577 @@
...
+            number of filters the rounding process potentially produces
+            (as per v1).
+        stem_conv_padding: str, can be 'same' or 'valid'. Padding for the stem.
+        bn_momentum: float, momentum for the moving average calcualtion in the
+            batch normalization layers.
+
+    Example:
+    ```python
+    # You can customize the EfficientNet architecture:
+    model = EfficientNetBackbone(
```
[the remark this answers is not in the document]
Comment (pkgoogle) [reply keras_nlp/src/models/efficientnet/efficientnet_backbone.py:88]: thanks!

File: keras_nlp/src/models/efficientnet/efficientnet_backbone.py [hunk 16]
```diff
@@ -0,0 +1,577 @@
...
+        stackwise_kernel_sizes=[3, 3, 3, 3, 3, 3],
+        stackwise_num_repeats=[2, 4, 4, 6, 9, 15],
+        stackwise_input_filters=[24, 24, 48, 64, 128, 160],
+        stackwise_output_filters=[24, 48, 64, 128, 160, 256],
+        stackwise_expansion_ratios=[1, 4, 4, 4, 6, 6],
+        stackwise_squeeze_and_excite_ratios=[0.0, 0.0, 0, 0.25, 0.25, 0.25],
+        stackwise_strides=[1, 2, 2, 2, 1, 2],
+        stackwise_conv_types=[
```
[the remark this answers is not in the document]
Comment (pkgoogle) [reply keras_nlp/src/models/efficientnet/efficientnet_backbone.py]: done

File: keras_nlp/src/models/efficientnet/efficientnet_backbone.py [hunk 17]
```diff
@@ -0,0 +1,577 @@
...
+                        inputs=x,
+                        filters_in=input_filters,
+                        filters_out=output_filters,
+                        kernel_size=stackwise_kernel_sizes[i],
+                        strides=strides,
+                        expand_ratio=stackwise_expansion_ratios[i],
+                        se_ratio=squeeze_and_excite_ratio,
+                        activation=activation,
+                        dropout_rate=drop_connect_rate * block_id / blocks,
+                        name="block{}{}_".format(i + 1, letter_identifier),
```
[the remark this answers is not in the document]
Comment (pkgoogle) [reply keras_nlp/src/models/efficientnet/efficientnet_backbone.py]: refactored to above and changed to f-string format.

File: keras_nlp/src/models/efficientnet/fusedmbconv.py [hunk 18]
```diff
@@ -0,0 +1,228 @@
...
+        "scale": 2.0,
+        "mode": "fan_out",
+        "distribution": "truncated_normal",
+    },
+}
+
+
+class FusedMBConvBlock(keras.layers.Layer):
+    """
+    Implementation of the FusedMBConv block (Fused Mobile Inverted Residual
```
[the remark this answers is not in the document]
Comment (pkgoogle) [reply keras_nlp/src/models/efficientnet/fusedmbconv.py]: done

File: keras_nlp/src/models/efficientnet/mbconv.py (hunk 6 above)
Comment (pkgoogle) [reply keras_nlp/src/models/efficientnet/mbconv.py]: done

**Review by mattdangerw — APPROVED** [review]
Thanks!
